[JIT] Expose `__deepcopy__` on script::Object

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40068 [JIT] Expose `__deepcopy__` on script::Object**

Differential Revision: [D22058808](https://our.internmc.facebook.com/intern/diff/D22058808)

Commit:
[JIT] Expose `__deepcopy__` on script::Object

[ghstack-poisoned]

Commit:
Update on "[JIT] Expose `__deepcopy__` on script::Object"


Differential Revision: [D22058808](https://our.internmc.facebook.com/intern/diff/D22058808)

[ghstack-poisoned]

Commit:
Update on "[JIT] Expose `__deepcopy__` on script::Object"


Differential Revision: [D22058808](https://our.internmc.facebook.com/intern/diff/D22058808)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -846,6 +846,12 @@ void initJitScriptBindings(PyObject* module) {
 
+  object_class.def(
+      "__deepcopy__", [](const Object& self, const py::dict& memo) {
+        // TODO: Figure out if memoization makes any sense in this context
```
It does: if two script objects exist as a separate part of the model but share state, this will break that sharing. It's probably not frequent, but it is one of those things that would be bad to leave waiting for someone to hit. If you stash the memo table for IValue's deep copy in the memo dictionary under some known key, then it can be used across deep copy calls in IValue.

Commit:
Update on "[JIT] Expose `__deepcopy__` on script::Object"


Differential Revision: [D22058808](https://our.internmc.facebook.com/intern/diff/D22058808)

[ghstack-poisoned]

jamesr66a commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -846,6 +846,22 @@ void initJitScriptBindings(PyObject* module) {
+  struct DeepCopyMemoTable {
+    std::shared_ptr<IValue::HashAliasedIValueMap> map;
+  };
+  py::class_<DeepCopyMemoTable>(m, "DeepCopyMemoTable");
```
@zdevito I couldn't figure out a better way to store the table in the map and get a mutable reference to it later via pybind so I added this indirection. (Issue was in `py::cast<HashedAliasedIValueMap>`, it claimed that the object was in the scope of the cast function or something)

zdevito replied:
This seems fine.

Commit:
Update on "[JIT] Expose `__deepcopy__` on script::Object"


Differential Revision: [D22058808](https://our.internmc.facebook.com/intern/diff/D22058808)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -846,6 +846,23 @@ void initJitScriptBindings(PyObject* module) {
+        }
+        auto& ivalue_memo =
+            *py::cast<DeepCopyMemoTable>(memo["__torch_script_memo_table"]).map;
+        return Object(IValue(self._ivalue()).deepcopy(ivalue_memo).toObject());
```
`self._ivallue()` is not an Object? seems like the IValue->Object thing is redundant.

jamesr66a replied:
no it's an `intrusive_ptr<ivalue::Object>`, not a `script::Object` :p

zdevito replied:
Does IValue not have a deepcopy method?

jamesr66a replied:
Yeah, it does. `ivalue::Object` doesn't. I think you're mis-reading the parens, `Object(` wraps the final value in an Object, not the result of `_ivalue()`

zdevito replied:
... and somehow toObject also doesn't return an object but an intrusive ptr to what used to be an object impl? The whole IValue and its members has gotten confusing.
